### DISCUSSION TITLE
set up for Supabase Auth with NextJS frontend (mainly server-side) and FastAPI backend?

### DISCUSSION THREAD

**Discussion by hsunami10** [supabase/supabase#28918] [Questions]
Hello! I’ve been doing some research and I cannot seem to find a way to use NextJS SSR with Supabase Auth + authenticate with a FastAPI backend server.

Most posts that I read refer to supabase’s auth client library — how you get the current session (in the browser) with `auth.getSession()`, then pass the access token in the `Authorization: Bearer` header, which makes total sense.

But, what I’m looking for is to be able to use supabase auth on the server side, using `auth.getUser()` because using `auth.getSession()` on the server side is **not recommended.**

Is it even possible to use `auth.getUser()` on NextJS server components to authenticate, and then somehow get the token to pass to FastAPI to authenticate, without using `auth.getSession()`?

**Comment by hsunami10** [comment]
basically, is there any way to only use supabase with NextJS, and not use the supabase-py library, and instead use a python-jwt or something?

**Comment by dimitarbarev** [comment]
Yes — FastAPI does not need supabase-py.

The important distinction is:

`getUser() / getClaims()` → validate the user's identity
`getSession()` → access the session itself, including the access token

Using `getSession()` server-side isn't inherently forbidden. The warning is that you shouldn't trust the user data from getSession() for authorization, because it isn't necessarily revalidated against Supabase Auth.

So in Next.js you can do:

```
const supabase = await createClient()

// Validate identity
const { data: { user }, error } = await supabase.auth.getUser()

if (error || !user) {
  // unauthorized
}

// Get the token that needs to be forwarded
const { data: { session } } = await supabase.auth.getSession()

await fetch("https://your-fastapi/api/whatever", {
  headers: {
    Authorization: `Bearer ${session!.access_token}`,
  },
})
```

Then FastAPI can independently verify that JWT. You don't need supabase-py for that.

With Supabase's asymmetric signing keys, FastAPI can verify the JWT using any good Python JWT library against Supabase's JWKS endpoint:

`https://<project-ref>.supabase.co/auth/v1/.well-known/jwks.json
`
Verify at least the signature, iss, aud and exp claims.

Architecturally:

Browser
  ↓ Supabase session cookie
Next.js
  ↓ validates user
  ↓ Authorization: Bearer <access_token>
FastAPI
  ↓ independently verifies JWT
Protected endpoint

So `getUser()` doesn't return an access token because that's not its purpose. If Next.js needs to forward the existing access token, getting it from the session is appropriate; just don't treat getSession().user itself as proof of identity.

Also, with current Supabase versions I'd use getClaims() instead of getUser() when you only need verified JWT claims, since it can verify the token against the project's JWKS without making an Auth server request each time.
